tests/ada_c.cpp

```diff
@@ -258,3 +258,86 @@ TEST(ada_c, ada_get_scheme_type) {
+                        value2.length());
+  ASSERT_EQ(ada_search_params_size(out), 5);
+
+  ASSERT_TRUE(ada_search_params_has(out, key.c_str(), key.length()));
+  ASSERT_FALSE(ada_search_params_has_value(out, key.c_str(), key.length(),
+                                           value.c_str(), value.length()));
+  ASSERT_TRUE(ada_search_params_has_value(out, key.c_str(), key.length(),
+                                          value2.c_str(), value2.length()));
+
+  ada_strings result =
+      ada_search_params_get_all(out, key.c_str(), key.length());
+  ASSERT_EQ(ada_strings_size(result), 1);
+  ada_strings_free(result);
+
+  ada_url_search_params_keys_iter keys = ada_search_params_get_keys(out);
+  ada_url_search_params_values_iter values = ada_search_params_get_values(out);
+  ada_url_search_params_entries_iter entries =
+      ada_search_params_get_entries(out);
+
+  ASSERT_FALSE(ada_search_params_keys_iter_is_done(keys));
+  ASSERT_FALSE(ada_search_params_values_iter_is_done(values));
+  ASSERT_FALSE(ada_search_params_entries_iter_is_done(entries));
+
+  ASSERT_EQ(convert_string(ada_search_params_keys_iter_next(keys)), "a");
+  ASSERT_EQ(convert_string(ada_search_params_keys_iter_next(keys)), "c");
+  ASSERT_EQ(convert_string(ada_search_params_keys_iter_next(keys)), "c");
+  ASSERT_EQ(convert_string(ada_search_params_keys_iter_next(keys)), "f");
+  ASSERT_EQ(convert_string(ada_search_params_keys_iter_next(keys)), "key1");
+  ASSERT_TRUE(ada_search_params_keys_iter_is_done(keys));
+
+  ASSERT_EQ(convert_string(ada_search_params_values_iter_next(values)), "b");
+  ASSERT_EQ(convert_string(ada_search_params_values_iter_next(values)), "d");
+  ASSERT_EQ(convert_string(ada_search_params_values_iter_next(values)), "e");
+  ASSERT_EQ(convert_string(ada_search_params_values_iter_next(values)), "g");
+  ASSERT_EQ(convert_string(ada_search_params_values_iter_next(values)),
+            "value2");
+  ASSERT_TRUE(ada_search_params_values_iter_is_done(values));
+
+  ada_string_pair pair = ada_search_params_entries_iter_next(entries);
+  ASSERT_EQ(convert_string(pair.value), "b");
+  ASSERT_EQ(convert_string(pair.key), "a");
+
+  pair = ada_search_params_entries_iter_next(entries);
+  ASSERT_EQ(convert_string(pair.value), "d");
+  ASSERT_EQ(convert_string(pair.key), "c");
+
+  while (!ada_search_params_entries_iter_is_done(entries)) {
+    ada_search_params_entries_iter_next(entries);
+  }
+
+  ada_search_params_remove(out, key.c_str(), key.length());
+  ada_search_params_remove_value(out, key.c_str(), key.length(), value.c_str(),
+                                 value.length());
+
+  ada_owned_string str = ada_search_params_to_string(out);
+  ASSERT_EQ(convert_string(str), "a=b&c=d&c=e&f=g");
+
+  ada_search_params_keys_iter_free(keys);
+  ada_search_params_values_iter_free(values);
+  ada_search_params_entries_iter_free(entries);
+  ada_free_owned_string(str);
+  ada_search_params_free(out);
```

Comment:
Since we have `ada_free_string` and `ada_free_owned_string`, it will be better to use `ada_free_search_params` (and also for iter frees)

```suggestion
  ada_free_search_params(out);
```